runtime/src/slots.rs

```diff
@@ -0,0 +1,1425 @@
+		/// parachain gets off-boarded; i.e. its lease period is up and it isn't renewed.
+		pub Offboarding get(offboarding): map ParaIdOf<T> => T::AccountId;
+	}
+}
+
+decl_event!(
+	pub enum Event<T> where
+		AccountId = <T as system::Trait>::AccountId,
+		BlockNumber = <T as system::Trait>::BlockNumber,
+		LeasePeriod = LeasePeriodOf<T>,
+		ParaId = ParaIdOf<T>,
+		Balance = BalanceOf<T>,
+	{
+		/// A new lease period is beginning.
+		NewLeasePeriod(LeasePeriod),
+		/// An auction started. Provides its index and the block number where it will begin to
+		/// close and the first lease period of the quadruplet that is auctioned.
+		AuctionStarted(AuctionIndex, LeasePeriod, BlockNumber),
+		/// An auction ended. All funds become unreserved.
+		AuctionClosed(AuctionIndex),
+		/// Someone won the right to deploy a parachain. Balance amount is deducted for deposit.
+		WonDeploy(NewBidder<AccountId>, SlotRange, ParaId, Balance),
+		/// An existing parachain won the right to continue.
+		WonRenewal(ParaId, SlotRange, Balance, Balance),
+		/// Funds were reserved for a winning bid. First balance is the extra amount reserved.
+		/// Second is the total.
+		Reserved(AccountId, Balance, Balance),
+		/// Funds were unreserved since bidder is no longer active.
+		Unreserved(AccountId, Balance),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event<T>() = default;
+
+		fn on_initialize(n: T::BlockNumber) {
+			let lease_period = T::LeasePeriod::get();
+			let lease_period_index: LeasePeriodOf<T> = (n / lease_period).into();
+
+			// Check to see if an auction just ended.
+			if let Some((winning_ranges, auction_lease_period_index)) = Self::check_auction_end(n) {
+				// Auction is ended now. We have the winning ranges and the lease period index which
+				// acts as the offset. Handle it.
+				Self::manage_auction_end(
+					lease_period_index,
+					auction_lease_period_index,
+					winning_ranges,
+				);
+			}
+			// If we're beginning a new lease period then handle that, too.
+			if (n % lease_period).is_zero() {
+				Self::manage_lease_period_start(lease_period_index);
+			}
+		}
+
+		fn on_finalize(now: T::BlockNumber) {
+			// If the current auction is in it ending period, then ensure that the (sub-)range
+			// winner information is duplicated from the previous block in case no bids happened
+			// in this block.
+			if let Some(offset) = Self::is_ending(now) {
+				if !<Winning<T>>::exists(&offset) {
+					<Winning<T>>::insert(offset,
+						offset.checked_sub(&One::one())
+							.and_then(<Winning<T>>::get)
+							.unwrap_or_default()
+					);
+				}
+			}
+		}
+
+		/// Create a new auction.
+		///
+		/// This can only happen when there isn't already an auction in progress and may only be
+		/// called by the root origin. Accepts the `duration` of this auction and the
+		/// `lease_period_index` at which it should begin.
```

Comment:
This is not written clearly to me. It sounds like the auction should begin at some `lease_period_index`.

>`lease_period_index` at which it should begin

I would write something like:

> Creates a new auction for some `lease_period` which starts now and ends after some `duration`.